File: python/src/prodcycle/api_client.py
"""ProdCycle Compliance API client.

Thin HTTP client for https://api.prodcycle.com/v1/compliance/* — no
proprietary policies or scanner code is bundled in the SDK; everything
runs server-side.

Mirrors node/src/api-client.ts so the two SDKs stay feature-equivalent.
"""

import json
import math
import os
import random
import sys
import time
import urllib.error
import urllib.parse
import urllib.request


DEFAULT_API_URL = "https://api.prodcycle.com"


def _env_int(name, fallback):
    """Read a positive integer from an env var or fall back to a default.

    Used for the timeout / retry knobs below so operators can tune the
    client in CI without forking the SDK.
    """
    raw = os.environ.get(name)
    if not raw:
        return fallback
    try:
        parsed = int(raw)
    except ValueError:
        return fallback
    return parsed if parsed > 0 else fallback


# Maximum retry attempts for 429/503 responses. After this many tries we
# give up and surface the error to the caller.
MAX_RETRY_ATTEMPTS = _env_int("PC_MAX_RETRY_ATTEMPTS", 4)

# Hard ceiling on Retry-After (seconds). Even if the server asks for more
# than this we cap it so the CLI doesn't appear to hang indefinitely on a
# misconfigured server.
MAX_RETRY_AFTER_SECONDS = _env_int("PC_MAX_RETRY_AFTER_SECONDS", 300)

# Per-request socket timeout. Without this `urllib.request.urlopen` would
# block indefinitely on a stalled server, bypassing the retry cap and the
# async-poll deadline. Default is 2 minutes — long enough for the largest
# non-async sync `/validate` call.
REQUEST_TIMEOUT_S = _env_int("PC_REQUEST_TIMEOUT_S", 120)

# Conservative client-side chunk sizing for the chunked-session flow.
# /chunks accepts up to 50 MB / 2000 files per request, but smaller chunks
# shorten tail-latency on a single saturated chunk; the server's per-content
# findings cache keeps re-scans of unchanged files O(1) regardless of chunk
# size, so picking on the smaller side costs little.
DEFAULT_CHUNK_MAX_BYTES = _env_int("PC_DEFAULT_CHUNK_MAX_BYTES", 5 * 1024 * 1024)
DEFAULT_CHUNK_MAX_FILES = _env_int("PC_DEFAULT_CHUNK_MAX_FILES", 200)

# Async-validate poll cadence.
ASYNC_POLL_INTERVAL_S = _env_int("PC_ASYNC_POLL_INTERVAL_S", 2)
ASYNC_POLL_TIMEOUT_S = _env_int("PC_ASYNC_POLL_TIMEOUT_S", 10 * 60)

# Keys in `options['config']` that route client-side behavior (sync / async
# / chunked, chunk sizing) and MUST NOT be forwarded to the server. The
# server's `options` schema rejects unknown keys strictly on some endpoints,
# so leaking these would cause 400s rather than just being inert.
_CLIENT_ONLY_CONFIG_KEYS = frozenset({"mode", "chunkMaxBytes", "chunkMaxFiles"})


class ApiError(Exception):
    """Raised for any non-2xx API response.

    Carries the parsed error body + status so callers can branch on
    ``body['error']['details']['suggestedEndpoint']`` (413 → chunked-session
    fallback) or ``retry_after_seconds`` (429 / 503 → backoff + retry).
    """

    def __init__(self, status_code, body, retry_after_seconds, message):
        super().__init__(message)
        self.status_code = status_code
        self.body = body
        self.retry_after_seconds = retry_after_seconds


class ComplianceApiClient:
    def __init__(self, api_url=None, api_key=None):
        self.api_url = (
            (api_url or os.environ.get("PC_API_URL") or DEFAULT_API_URL).rstrip("/")
        )
        self.api_key = api_key or os.environ.get("PC_API_KEY", "")

        if (
            not self.api_key
            and os.environ.get("PYTEST_CURRENT_TEST") is None
            and not os.environ.get("PC_SUPPRESS_WARNINGS")
        ):
            sys.stderr.write(
                "Warning: PC_API_KEY is not set. API calls will likely fail.\n"
            )

    # ─── Synchronous validate ────────────────────────────────────────────

    def validate(self, files, frameworks, options=None):
        """Synchronous validate.

        On a 413 with ``details.suggestedEndpoint='/v1/compliance/scans'``,
        silently falls back to the chunked-session flow so large-repo CI
        jobs don't have to know the difference.
        """
        opts_payload = self._build_options(options)
        try:
            return self._request(
                "POST",
                "/v1/compliance/validate",
                {"files": files, "frameworks": frameworks, "options": opts_payload},
            )
        except ApiError as err:
            if (
                err.status_code == 413
                and err.body
                and err.body.get("error", {})
                .get("details", {})
                .get("suggestedEndpoint")
                == "/v1/compliance/scans"
            ):
                # Server says: this payload won't fit, use chunked sessions
                # instead. Fall back transparently — the caller asked for
                # validate(), the semantics (single scanId with final
                # findings) are preserved.
                return self.validate_chunked(files, frameworks, options)
            raise

    def hook(self, files, frameworks, options=None):
        """Hook endpoint — small per-write call from coding agents.

        No suggestedEndpoint fallback because /hook keeps the historical
        50 MB ceiling; if a single hook write exceeds that, the caller's
        batching is the bug to fix.
        """
        return self._request(
            "POST",
            "/v1/compliance/hook",
            {
                "files": files,
                "frameworks": frameworks,
                "options": self._build_options(options),
            },
        )

    # ─── Chunked sessions ────────────────────────────────────────────────

    def open_session(self, frameworks, options=None):
        """Open a chunked scan session.

        Returns a ``scanId`` that subsequent ``append_chunk`` /
        ``complete_session`` calls reference. Server-side TTL is 30 min
        by default — abandoned sessions self-clean via the
        stale-session reaper.
        """
        return self._request(
            "POST",
            "/v1/compliance/scans",
            {
                "frameworks": frameworks,
                "options": self._build_options(options),
            },
        )

    def append_chunk(self, scan_id, files):
        """Append a chunk of files to an open session.

        Each call has its own /hook-style cap (50 MB / 2000 files). The
        server caches per-content findings, so re-scans of unchanged files
        are O(1).
        """
        return self._request(
            "POST",
            f"/v1/compliance/scans/{urllib.parse.quote(scan_id, safe='')}/chunks",
            {"files": files},
        )

    def complete_session(self, scan_id):
        """Finalize a chunked session.

        Flips status to COMPLETED, computes summary + passed, returns
        final findings.
        """
        return self._request(
            "POST",
            f"/v1/compliance/scans/{urllib.parse.quote(scan_id, safe='')}/complete",
            {},
        )

    def validate_chunked(self, files, frameworks, options=None):
        """High-level helper: open → append (in chunks) → complete.

        Returns the same shape as ``validate()`` so callers that
        auto-fallback don't have to special-case the result.
        """
        options = options or {}
        config = options.get("config") or {}
        chunk_max_bytes = int(config.get("chunkMaxBytes", DEFAULT_CHUNK_MAX_BYTES))
        chunk_max_files = int(config.get("chunkMaxFiles", DEFAULT_CHUNK_MAX_FILES))

        session = self.open_session(frameworks, options)
        scan_id = session["scanId"]

        for chunk in chunk_files(files, chunk_max_bytes, chunk_max_files):
            self.append_chunk(scan_id, chunk)

        result = self.complete_session(scan_id)
        # Preserve the scanId even if the server's complete response doesn't
        # echo it back, so callers always have it in hand.
        if "scanId" not in result:
            result["scanId"] = scan_id
        return result

    # ─── Async validate ──────────────────────────────────────────────────

    def validate_async(self, files, frameworks, options=None):
        """Async-validate: returns a scanId immediately.

        Caller polls ``get_scan(scanId)`` until status is COMPLETED or
        FAILED. Useful for CI runners that don't want to hold a connection
        for a 60 s scan.

        On a 413 with ``details.suggestedEndpoint='/v1/compliance/scans'``,
        transparently falls back to the chunked-session flow — same
        contract as ``validate()``. ``validate_chunked()`` returns the
        full final result rather than a kickoff envelope, so callers
        that branch on ``status='IN_PROGRESS'`` should treat a chunked
        return as a fully-completed scan.
        """
        try:
            return self._request(
                "POST",
                "/v1/compliance/validate?async=true",
                {
                    "files": files,
                    "frameworks": frameworks,
                    "options": self._build_options(options),
                },
            )
        except ApiError as err:
            if (
                err.status_code == 413
                and isinstance(err.body, dict)
                and err.body.get("error", {})
                .get("details", {})
                .get("suggestedEndpoint")
                == "/v1/compliance/scans"
            ):
                return self.validate_chunked(files, frameworks, options)
            raise

    def get_scan(self, scan_id):
        """Fetch the current state of any scan (sync, async, chunked)."""
        return self._request(
            "GET",
            f"/v1/compliance/scans/{urllib.parse.quote(scan_id, safe='')}",
            None,
        )

    def validate_and_poll(self, files, frameworks, options=None):
        """High-level helper: kick off async-validate, poll until terminal.

        Returns the same shape as ``validate()``.
        """
        kickoff = self.validate_async(files, frameworks, options)
        scan_id = kickoff["scanId"]
        # validate_async() may have transparently fallen back to the
        # chunked-session flow on a 413 — in which case `kickoff` is
        # already a fully-completed scan, not just a kickoff envelope.
        # Short-circuit so we don't burn an extra get_scan() round-trip
        # against a scan that's already terminal.
        if kickoff.get("status") in ("COMPLETED", "FAILED", "PASSED"):
            return kickoff
        deadline = time.monotonic() + ASYNC_POLL_TIMEOUT_S

        # Always poll at least once, and re-check the deadline AFTER the
        # blocking get_scan() returns rather than only at loop top — a
        # single stalled get_scan() could otherwise extend wall-clock
        # well past ASYNC_POLL_TIMEOUT_S (a stuck call sleeps up to
        # REQUEST_TIMEOUT_S on its own). Belt-and-suspenders: also check
        # before sleeping so a scan that completes during the trailing
        # sleep window isn't reported as a timeout.
        # Match the terminal-status set used by the short-circuit above
        # — some endpoints/older API responses surface 'PASSED' rather
        # than 'COMPLETED' for a successful scan. Treat all three as
        # terminal so the loop doesn't spin out to the timeout while the
        # scan is actually finished.
        while True:
            scan = self.get_scan(scan_id)
            if scan.get("status") in ("COMPLETED", "FAILED", "PASSED"):
                if "scanId" not in scan:
                    scan["scanId"] = scan_id
                return scan
            if time.monotonic() >= deadline:
                break
            time.sleep(ASYNC_POLL_INTERVAL_S)

        raise Exception(
            f"Async validate scan {scan_id} did not complete within "
            f"{ASYNC_POLL_TIMEOUT_S}s. Re-run with the same scanId to keep "
            f"polling: prodcycle scans {scan_id}"
        )

    # ─── Internals ───────────────────────────────────────────────────────

    def _build_options(self, options):
        options = options or {}
        opts_payload = {
            "severity_threshold": options.get("severityThreshold"),
            "fail_on": options.get("failOn"),
        }
        config = options.get("config")
        if isinstance(config, dict):
            # Strip client-routing keys (mode / chunkMaxBytes / chunkMaxFiles)
            # before forwarding — they steer this SDK, not the server, and
            # leaking them produces 400s on endpoints with strict schemas.
            opts_payload.update(
                {k: v for k, v in config.items() if k not in _CLIENT_ONLY_CONFIG_KEYS}
            )
        # Drop None values — server treats absent + null differently for
        # some keys (severity_threshold default is 'low'; absent uses
        # the workspace default).
        return {k: v for k, v in opts_payload.items() if v is not None}

    def _request(self, method, endpoint, data):
        """Single HTTP request with auth, retry, and structured errors."""
        url = f"{self.api_url}{endpoint}"
        last_error = None

        for attempt in range(MAX_RETRY_ATTEMPTS):
            req = urllib.request.Request(url, method=method)
            req.add_header("Authorization", f"Bearer {self.api_key}")
            if method == "POST":
                req.add_header("Content-Type", "application/json")

            payload = json.dumps(data).encode("utf-8") if data is not None else None

            try:
                with urllib.request.urlopen(
                    req, data=payload, timeout=REQUEST_TIMEOUT_S
                ) as response:
                    body = response.read().decode("utf-8")
                    parsed = json.loads(body) if body else {}
                    return _unwrap_envelope(parsed)
            except urllib.error.HTTPError as err:
                err_body_raw = ""
                try:
                    err_body_raw = err.read().decode("utf-8")
                except Exception:
                    pass
                try:
                    err_body = json.loads(err_body_raw) if err_body_raw else None
                except Exception:
                    err_body = None

                retry_after = _parse_retry_after(err.headers.get("Retry-After"))
                message = (
                    (err_body or {}).get("error", {}).get("message")
                    or f"API request failed with status {err.code}"
                )

                is_retryable = err.code in (429, 503)
                if is_retryable and attempt < MAX_RETRY_ATTEMPTS - 1:
                    delay = (
                        retry_after
                        if retry_after is not None
                        else math.ceil(_retry_backoff_ms(attempt) / 1000)
                    )
                    time.sleep(min(delay, MAX_RETRY_AFTER_SECONDS))
                    continue

                raise ApiError(err.code, err_body, retry_after, message)
            except urllib.error.URLError as err:
                # Connection-level failures (DNS, TCP, TLS). Retryable
                # up to the same cap as 503 — the server may be
                # momentarily down or the network blip may resolve.
                last_error = err
                if attempt < MAX_RETRY_ATTEMPTS - 1:
                    time.sleep(_retry_backoff_ms(attempt) / 1000.0)
                    continue
                raise Exception(
                    f"Failed to connect to ProdCycle API: {err.reason}"
                )

        # Unreachable in practice: every iteration returns, raises, or
        # continues. Kept solely so static analysis can tell that the
        # function always returns or raises (mypy / type-checkers).
        raise AssertionError(  # pragma: no cover
            f"_request loop exited without returning or raising "
            f"(last_error={last_error})"
        )


# ─── Helpers ─────────────────────────────────────────────────────────────


def _unwrap_envelope(parsed):
    """Unwrap ``{status, statusCode, data}`` envelope if present.

    Older/local deployments may return the bare body — fall through in
    that case.
    """
    if (
        isinstance(parsed, dict)
        and "data" in parsed
        and isinstance(parsed.get("data"), dict)
        and "status" in parsed
    ):
        return parsed["data"]
    return parsed


def _retry_backoff_ms(attempt):
    """Exponential backoff with jitter (1s, 2s, 4s, 8s, ...)."""
    base = 1000 * (2 ** attempt)
    jitter = random.random() * 500
    return base + jitter


def _parse_retry_after(value):
    """Parse Retry-After header. Returns seconds (int) or None."""
    if not value:
        return None
    try:
        return max(0, int(value))
    except ValueError:
        pass
    # HTTP-date form: 'Sun, 06 Nov 1994 08:49:37 GMT'
    try:
        from email.utils import parsedate_to_datetime

        dt = parsedate_to_datetime(value)
        return max(0, math.ceil(dt.timestamp() - time.time()))
    except Exception:
        return None


def chunk_files(files, max_bytes, max_files):
    """Split ``{path: content}`` into chunks under both byte + file caps.

    UTF-8 byte-length is used since the server counts the request body's
    bytes after JSON serialisation; this is a conservative client-side
    approximation.
    """
    chunks = []
    current = {}
    current_bytes = 0
    current_count = 0

    for file_path, content in files.items():
        file_bytes = len(content.encode("utf-8")) + len(file_path.encode("utf-8"))
        # Single file exceeds the cap on its own — emit it as its own
        # chunk and let the server's per-file cap reject if needed.
        # Common case: huge SQL dumps, generated bundles.
        if file_bytes > max_bytes:
            if current_count > 0:
                chunks.append(current)
                current = {}
                current_bytes = 0
                current_count = 0
            chunks.append({file_path: content})
            continue
        if (
            current_bytes + file_bytes > max_bytes
            or current_count + 1 > max_files
        ):
            chunks.append(current)
            current = {}
            current_bytes = 0
            current_count = 0
        current[file_path] = content
        current_bytes += file_bytes
        current_count += 1

    if current_count > 0:
        chunks.append(current)

    return chunks

File: node/src/api-client.ts
export interface ScanOptions {
  severityThreshold?: 'low' | 'medium' | 'high' | 'critical';
  failOn?: ('low' | 'medium' | 'high' | 'critical')[];
  include?: string[];
  exclude?: string[];
  apiKey?: string;
  apiUrl?: string;
  config?: Record<string, unknown>;
}

export interface GateOptions {
  files: Record<string, string>;
  frameworks?: string[];
  severityThreshold?: 'low' | 'medium' | 'high' | 'critical';
  failOn?: ('low' | 'medium' | 'high' | 'critical')[];
  apiKey?: string;
  apiUrl?: string;
  config?: Record<string, unknown>;
}

export interface ScanResult {
  scanId?: string;
  passed: boolean;
  findingsCount?: number;
  findings?: unknown[];
  summary?: unknown;
  prompt?: string;
  status?: 'IN_PROGRESS' | 'COMPLETED' | 'FAILED';
  [key: string]: unknown;
}

interface ApiErrorBody {
  status: 'error';
  statusCode: number;
  error?: {
    type?: string;
    message?: string;
    suggestion?: string;
    details?: {
      maxBytes?: number;
      maxFiles?: number;
      chunkSizeBytes?: number;
      receivedBytes?: number;
      suggestedEndpoint?: string;
      [key: string]: unknown;
    };
  };
}

/**
 * Error thrown for any non-2xx response. Carries the parsed body + status so
 * callers can branch on `details.suggestedEndpoint` (413 → chunked-session
 * fallback) or `Retry-After` (429 / 503 → backoff + retry).
 */
export class ApiError extends Error {
  constructor(
    public readonly statusCode: number,
    public readonly body: ApiErrorBody | null,
    public readonly retryAfterSeconds: number | null,
    message: string,
  ) {
    super(message);
    this.name = 'ApiError';
  }
}

const DEFAULT_API_URL = 'https://api.prodcycle.com';

/**
 * Read a positive integer from an env var or fall back to a default. Used
 * for the timeout / retry knobs below so operators can tune behavior in CI
 * without forking the CLI.
 */
function envInt(name: string, fallback: number): number {
  const raw = process.env[name];
  if (!raw) return fallback;
  const parsed = Number(raw);
  return Number.isFinite(parsed) && parsed > 0 ? Math.floor(parsed) : fallback;
}

/**
 * Maximum retry attempts for 429/503 responses. After this many tries we
 * give up and surface the error to the caller.
 */
const MAX_RETRY_ATTEMPTS = envInt('PC_MAX_RETRY_ATTEMPTS', 4);

/**
 * Hard ceiling on Retry-After (seconds). Even if the server asks for more
 * than this we cap it so the CLI doesn't appear to hang indefinitely on a
 * misconfigured server.
 */
const MAX_RETRY_AFTER_SECONDS = envInt('PC_MAX_RETRY_AFTER_SECONDS', 300);

/**
 * Per-request fetch timeout. Without this a stalled connection would tie
 * up the CLI indefinitely, bypassing both the retry cap and the async-poll
 * deadline. Default is 2 minutes — long enough for the largest non-async
 * sync `/validate` call, short enough that a hung TCP socket gets aborted.
 */
const REQUEST_TIMEOUT_MS = envInt('PC_REQUEST_TIMEOUT_MS', 120_000);

/**
 * Conservative client-side chunk sizing for the chunked-session flow. The
 * /chunks endpoint accepts up to 50 MB / 2000 files per request, but most
 * customer payloads are well under this and smaller chunks shorten
 * tail-latency on a single saturated chunk. The server's per-content
 * findings cache means re-scans of unchanged files are O(1) regardless of
 * chunk size, so picking on the smaller side costs little.
 */
const DEFAULT_CHUNK_MAX_BYTES = envInt(
  'PC_DEFAULT_CHUNK_MAX_BYTES',
  5 * 1024 * 1024, // 5 MB
);
const DEFAULT_CHUNK_MAX_FILES = envInt('PC_DEFAULT_CHUNK_MAX_FILES', 200);

/**
 * Async-validate poll cadence. The server typically completes scans in
 * 10–60 s; polling every 2 s keeps the round-trip overhead bounded while
 * still feeling responsive in interactive use.
 */
const ASYNC_POLL_INTERVAL_MS = envInt('PC_ASYNC_POLL_INTERVAL_MS', 2000);
const ASYNC_POLL_TIMEOUT_MS = envInt(
  'PC_ASYNC_POLL_TIMEOUT_MS',
  10 * 60 * 1000, // 10 minutes
);

export class ComplianceApiClient {
  private apiUrl: string;
  private apiKey: string;

  constructor(apiUrl?: string, apiKey?: string) {
    this.apiUrl = apiUrl || process.env.PC_API_URL || DEFAULT_API_URL;
    this.apiKey = apiKey || process.env.PC_API_KEY || '';

    if (
      !this.apiKey &&
      process.env.NODE_ENV !== 'test' &&
      !process.env.PC_SUPPRESS_WARNINGS
    ) {
      process.stderr.write(
        'Warning: PC_API_KEY is not set. API calls will likely fail.\n',
      );
    }
  }

  /**
   * Synchronous validate. On a 413 with `details.suggestedEndpoint ===
   * '/v1/compliance/scans'`, silently falls back to the chunked-session
   * flow so large-repo CI jobs don't have to know the difference.
   */
  async validate(
    files: Record<string, string>,
    frameworks: string[],
    options: ScanOptions = {},
  ): Promise<ScanResult> {
    try {
      return await this.request('POST', '/v1/compliance/validate', {
        files,
        frameworks,
        options: this.buildOptions(options),
      });
    } catch (err) {
      if (
        err instanceof ApiError &&
        err.statusCode === 413 &&
        err.body?.error?.details?.suggestedEndpoint === '/v1/compliance/scans'
      ) {
        // Server says: this payload won't fit, use chunked sessions instead.
        // Fall back transparently — the caller asked for `validate`, the
        // semantics (single scanId with final findings) are preserved.
        return this.validateChunked(files, frameworks, options);
      }
      throw err;
    }
  }

  /**
   * Hook endpoint — small per-write call from coding agents. No
   * suggestedEndpoint fallback because /hook keeps the historical 50 MB
   * ceiling; if a single hook write exceeds that, the caller's batching
   * is the bug to fix.
   */
  async hook(
    files: Record<string, string>,
    frameworks: string[],
    options: ScanOptions = {},
  ): Promise<ScanResult> {
    return this.request('POST', '/v1/compliance/hook', {
      files,
      frameworks,
      options: this.buildOptions(options),
    });
  }

  // ─── Chunked sessions ───────────────────────────────────────────────────

  /**
   * Open a chunked scan session. Returns a `scanId` that subsequent
   * `appendChunk` / `completeSession` calls reference. Server-side TTL is
   * 30 minutes by default — abandoned sessions self-clean via the
   * stale-session reaper.
   */
  async openSession(
    frameworks: string[],
    options: ScanOptions = {},
  ): Promise<{ scanId: string; chunkSizeBytes: number; maxFilesPerChunk: number; expiresAt: string }> {
    return this.request('POST', '/v1/compliance/scans', {
      frameworks,
      options: this.buildOptions(options),
    });
  }

  /**
   * Append a chunk of files to an open session. Each call has its own
   * /hook-style cap (50 MB / 2000 files). The server caches per-content
   * findings, so re-scans of unchanged files are O(1).
   */
  async appendChunk(
    scanId: string,
    files: Record<string, string>,
  ): Promise<{ filesScanned: number; cachedFiles: number; findingsAdded: number }> {
    return this.request('POST', `/v1/compliance/scans/${encodeURIComponent(scanId)}/chunks`, {
      files,
    });
  }

  /**
   * Finalize a chunked session: flips status to COMPLETED, computes
   * summary + passed, returns final findings.
   */
  async completeSession(scanId: string): Promise<ScanResult> {
    return this.request(
      'POST',
      `/v1/compliance/scans/${encodeURIComponent(scanId)}/complete`,
      {},
    );
  }

  /**
   * High-level helper: open → append (in chunks) → complete. Returns the
   * same shape as `validate()` so callers that auto-fallback don't have
   * to special-case the result.
   *
   * Caller can pre-set `chunkMaxBytes` / `chunkMaxFiles` on `options.config`
   * to override the conservative defaults.
   */
  async validateChunked(
    files: Record<string, string>,
    frameworks: string[],
    options: ScanOptions = {},
  ): Promise<ScanResult> {
    const chunkMaxBytes =
      (options.config?.chunkMaxBytes as number | undefined) ?? DEFAULT_CHUNK_MAX_BYTES;
    const chunkMaxFiles =
      (options.config?.chunkMaxFiles as number | undefined) ?? DEFAULT_CHUNK_MAX_FILES;

    const session = await this.openSession(frameworks, options);

    const chunks = chunkFiles(files, chunkMaxBytes, chunkMaxFiles);
    for (const chunk of chunks) {
      await this.appendChunk(session.scanId, chunk);
    }

    const result = await this.completeSession(session.scanId);
    return { scanId: session.scanId, ...result };
  }

  // ─── Async validate ─────────────────────────────────────────────────────

  /**
   * Async-validate: returns a `scanId` immediately; caller polls
   * `getScan(scanId)` until status is COMPLETED or FAILED. Useful for CI
   * runners that don't want to hold a connection for a 60 s scan.
   */
  async validateAsync(
    files: Record<string, string>,
    frameworks: string[],
    options: ScanOptions = {},
  ): Promise<{ scanId: string }> {
    return this.request('POST', '/v1/compliance/validate?async=true', {
      files,
      frameworks,
      options: this.buildOptions(options),
    });
  }

  /**
   * Fetch the current state of any scan (sync, async, or chunked-session).
   */
  async getScan(scanId: string): Promise<ScanResult> {
    return this.request('GET', `/v1/compliance/scans/${encodeURIComponent(scanId)}`, null);
  }

  /**
   * High-level helper: kicks off an async-validate, polls until terminal,
   * returns the same shape as `validate()`.
   */
  async validateAndPoll(
    files: Record<string, string>,
    frameworks: string[],
    options: ScanOptions = {},
  ): Promise<ScanResult> {
    const { scanId } = await this.validateAsync(files, frameworks, options);
    const deadline = Date.now() + ASYNC_POLL_TIMEOUT_MS;

    while (Date.now() < deadline) {
      const scan = await this.getScan(scanId);
      if (scan.status === 'COMPLETED' || scan.status === 'FAILED') {
        return { scanId, ...scan };
      }
      await sleep(ASYNC_POLL_INTERVAL_MS);
    }

    throw new Error(
      `Async validate scan ${scanId} did not complete within ${
        ASYNC_POLL_TIMEOUT_MS / 1000
      }s. Re-run with the same scanId to keep polling: pc scans get ${scanId}`,
    );
  }

  // ─── Internals ──────────────────────────────────────────────────────────

  private buildOptions(options: ScanOptions): Record<string, unknown> {
    return {
      severity_threshold: options.severityThreshold,
      fail_on: options.failOn,
      ...options.config,
    };
  }

  /**
   * Single HTTP request with:
   *   - auth header
   *   - 429/503 + Retry-After honoring (up to MAX_RETRY_ATTEMPTS)
   *   - structured error with parsed body so callers can introspect
   *     `details.suggestedEndpoint` etc.
   */
  private async request<T = ScanResult>(
    method: 'GET' | 'POST',
    endpoint: string,
    data: unknown,
  ): Promise<T> {
    const url = `${this.apiUrl.replace(/\/+$/, '')}${endpoint}`;

    let lastError: Error | null = null;
    for (let attempt = 0; attempt < MAX_RETRY_ATTEMPTS; attempt++) {
      let response: Response;
      try {
        response = await fetch(url, {
          method,
          headers: {
            Authorization: `Bearer ${this.apiKey}`,
            ...(method === 'POST' ? { 'Content-Type': 'application/json' } : {}),
          },
          ...(data !== null ? { body: JSON.stringify(data) } : {}),
          signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
        });
      } catch (networkErr: unknown) {
        // Connection-level failures (DNS, TCP, TLS). Treat as retryable up
        // to the same cap as 503 — the server may be momentarily down or
        // the network blip may resolve.
        lastError =
          networkErr instanceof Error ? networkErr : new Error(String(networkErr));
        if (attempt < MAX_RETRY_ATTEMPTS - 1) {
          await sleep(retryBackoffMs(attempt));
          continue;
        }
        throw new Error(`Failed to connect to ProdCycle API: ${lastError.message}`);
      }

      const responseText = await response.text();
      let parsed: unknown = null;
      try {
        parsed = responseText ? JSON.parse(responseText) : null;
      } catch {
        // Non-JSON body (e.g. ALB-level 502/504). Leave parsed as null;
        // the retry path below handles based on status code.
      }

      if (response.ok) {
        // Unwrap {status, statusCode, data: {...}} envelope if present.
        if (
          parsed &&
          typeof parsed === 'object' &&
          'data' in parsed &&
          (parsed as { data: unknown }).data &&
          typeof (parsed as { data: unknown }).data === 'object' &&
          'status' in parsed
        ) {
          return (parsed as { data: T }).data;
        }
        return (parsed as T) ?? ({} as T);
      }

      // Non-2xx. Inspect the body + Retry-After to decide whether to retry.
      const retryAfterSeconds = parseRetryAfter(response.headers.get('retry-after'));
      const errorBody = (parsed as ApiErrorBody | null) ?? null;
      const errorMessage =
        errorBody?.error?.message ?? `API request failed with status ${response.status}`;

      const isRetryable = response.status === 429 || response.status === 503;
      if (isRetryable && attempt < MAX_RETRY_ATTEMPTS - 1) {
        const delayMs =
          retryAfterSeconds != null ? retryAfterSeconds * 1000 : retryBackoffMs(attempt);
        const cappedDelayMs = Math.min(delayMs, MAX_RETRY_AFTER_SECONDS * 1000);
        await sleep(cappedDelayMs);
        continue;
      }

      throw new ApiError(response.status, errorBody, retryAfterSeconds, errorMessage);
    }

    // Loop exited via continue with no successful response — surface the
    // last error rather than returning silently.
    throw lastError ?? new Error('Exhausted retries without a response');
  }
}

// ─── Helpers ──────────────────────────────────────────────────────────────

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Exponential backoff with jitter for retryable errors that don't carry an
 * explicit Retry-After (network failures, malformed 503).
 */
function retryBackoffMs(attempt: number): number {
  const base = 1000 * 2 ** attempt; // 1s, 2s, 4s, 8s, ...
  const jitter = Math.random() * 500;
  return base + jitter;
}

/**
 * Parse Retry-After header. Spec allows either:
 *   - delta-seconds (an integer)
 *   - HTTP-date
 * We support both. Returns seconds as a non-negative integer, or null if
 * the header is missing/unparseable.
 */
function parseRetryAfter(value: string | null): number | null {
  if (!value) return null;
  const asInt = Number.parseInt(value, 10);
  if (!Number.isNaN(asInt)) return Math.max(0, asInt);
  const asDate = Date.parse(value);
  if (!Number.isNaN(asDate)) {
    return Math.max(0, Math.ceil((asDate - Date.now()) / 1000));
  }
  return null;
}

/**
 * Split a `{ path: content }` map into chunks that respect both a byte
 * cap and a file-count cap. UTF-8 byte-length is used since the server
 * counts the request body's bytes after JSON serialisation; this is a
 * conservative client-side approximation.
 */
export function chunkFiles(
  files: Record<string, string>,
  maxBytes: number,
  maxFiles: number,
): Record<string, string>[] {
  const chunks: Record<string, string>[] = [];
  let current: Record<string, string> = {};
  let currentBytes = 0;
  let currentCount = 0;

  for (const [filePath, content] of Object.entries(files)) {
    const fileBytes = Buffer.byteLength(content, 'utf8') + Buffer.byteLength(filePath, 'utf8');
    // If a single file exceeds the cap on its own we can't split it further
    // here — emit it as its own chunk and let the server's per-file cap (if
    // any) reject if needed. Common case: huge SQL dumps, generated bundles.
    if (fileBytes > maxBytes) {
      if (currentCount > 0) {
        chunks.push(current);
        current = {};
        currentBytes = 0;
        currentCount = 0;
      }
      chunks.push({ [filePath]: content });
      continue;
    }
    if (currentBytes + fileBytes > maxBytes || currentCount + 1 > maxFiles) {
      chunks.push(current);
      current = {};
      currentBytes = 0;
      currentCount = 0;
    }
    current[filePath] = content;
    currentBytes += fileBytes;
    currentCount += 1;
  }

  if (currentCount > 0) {
    chunks.push(current);
  }

  return chunks;
}
